File: src/qb2api/anthropic.py
"""Anthropic Messages API compatibility helpers."""

import json
import uuid
from typing import Any

from .anthropic_content import anthropic_message_to_openai as _anthropic_message_to_openai
from .anthropic_content import content_to_text as _content_to_text


def anthropic_to_openai(body: dict[str, Any]) -> dict[str, Any]:
    """Convert an Anthropic Messages request body into an OpenAI chat request."""
    request = _request_base(body)
    request.update(_optional_request_values(body))
    return request


def _request_base(body: dict[str, Any]) -> dict[str, Any]:
    messages: list[dict[str, Any]] = []
    system_text = _content_to_text(body.get("system"))
    if system_text:
        messages.append({"role": "system", "content": system_text})
    for message in body.get("messages", []):
        messages.extend(_anthropic_message_to_openai(message))
    return {
        "model": body.get("model"),
        "messages": messages,
        "stream": bool(body.get("stream", False)),
    }


def _optional_request_values(body: dict[str, Any]) -> dict[str, Any]:
    request: dict[str, Any] = {}
    for values in (
        _completion_options(body),
        _metadata_options(body),
        _tool_options(body),
        _context_options(body),
    ):
        request.update(values)
    return request


def _completion_options(body: dict[str, Any]) -> dict[str, Any]:
    request: dict[str, Any] = {}
    if body.get("max_tokens") is not None:
        request["max_tokens"] = body["max_tokens"]
    for source, target in (
        ("temperature", "temperature"),
        ("top_p", "top_p"),
        ("stop_sequences", "stop"),
    ):
        if body.get(source) is not None:
            request[target] = body[source]
    return request


def _metadata_options(body: dict[str, Any]) -> dict[str, Any]:
    metadata = body.get("metadata") or {}
    if isinstance(metadata, dict) and metadata.get("user_id"):
        return {"user": metadata["user_id"]}
    return {}


def _tool_options(body: dict[str, Any]) -> dict[str, Any]:
    request: dict[str, Any] = {}
    tools = [_anthropic_tool_to_openai(tool) for tool in body.get("tools") or []]
    if tools:
        request["tools"] = tools
    tool_choice = _anthropic_tool_choice_to_openai(body.get("tool_choice"))
    if tool_choice is not None:
        request["tool_choice"] = tool_choice
    return request


def _context_options(body: dict[str, Any]) -> dict[str, Any]:
    request: dict[str, Any] = {}
    for key in ("reasoning_effort", "context_window", "max_context_tokens"):
        if body.get(key) is not None:
            request[key] = body[key]
    return request


def openai_to_anthropic(response: dict[str, Any], model: str) -> dict[str, Any]:
    """Convert an OpenAI chat completion response into Anthropic message shape."""
    choice = (response.get("choices") or [{}])[0]
    message = choice.get("message") or {}
    return _anthropic_response(
        response=response,
        choice=choice,
        message=message,
        model=model,
    )


def _anthropic_response(
    *,
    response: dict[str, Any],
    choice: dict[str, Any],
    message: dict[str, Any],
    model: str,
) -> dict[str, Any]:
    return {
        "id": _message_id(response.get("id")),
        "type": "message",
        "role": "assistant",
        "model": model,
        "content": _response_content(message),
        "stop_reason": map_finish_reason(choice.get("finish_reason")),
        "stop_sequence": None,
        "usage": _response_usage(response),
    }


def _response_content(message: dict[str, Any]) -> list[dict[str, Any]]:
    content: list[dict[str, Any]] = []
    text = message.get("content")
    if text:
        content.append({"type": "text", "text": str(text)})
    for tool_call in message.get("tool_calls") or []:
        content.append(_tool_use_content(tool_call))
    return content


def _tool_use_content(tool_call: dict[str, Any]) -> dict[str, Any]:
    function = tool_call.get("function") or {}
    return {
        "type": "tool_use",
        "id": tool_call.get("id") or f"toolu_{uuid.uuid4().hex[:16]}",
        "name": function.get("name") or "",
        "input": _parse_tool_input(function.get("arguments")),
    }


def _response_usage(response: dict[str, Any]) -> dict[str, int]:
    usage = response.get("usage") or {}
    return {
        "input_tokens": int(usage.get("prompt_tokens") or usage.get("input_tokens") or 0),
        "output_tokens": int(usage.get("completion_tokens") or usage.get("output_tokens") or 0),
    }


def _anthropic_tool_to_openai(tool: dict[str, Any]) -> dict[str, Any]:
    return {
        "type": "function",
        "function": {
            "name": tool.get("name") or "",
            "description": tool.get("description") or "",
            "parameters": tool.get("input_schema") or {"type": "object", "properties": {}},
        },
    }


def _anthropic_tool_choice_to_openai(tool_choice: Any) -> Any:
    if tool_choice is None:
        return None
    if isinstance(tool_choice, str):
        return tool_choice
    if not isinstance(tool_choice, dict):
        return tool_choice
    choice_type = tool_choice.get("type")
    if choice_type == "auto":
        return "auto"
    if choice_type == "any":
        return "required"
    if choice_type == "none":
        return "none"
    if choice_type == "tool":
        return {"type": "function", "function": {"name": tool_choice.get("name") or ""}}
    return tool_choice


def _parse_tool_input(arguments: Any) -> dict[str, Any]:
    if isinstance(arguments, dict):
        return arguments
    if not arguments:
        return {}
    try:
        parsed = json.loads(str(arguments))
    except json.JSONDecodeError:
        return {"_raw": str(arguments)}
    return parsed if isinstance(parsed, dict) else {"value": parsed}


def _message_id(openai_id: Any) -> str:
    raw = str(openai_id or "")
    if raw.startswith("msg_"):
        return raw
    suffix = raw.removeprefix("chatcmpl-") or uuid.uuid4().hex[:24]
    return f"msg_{suffix}"


def map_finish_reason(reason: Any) -> str | None:
    if reason is None:
        return None
    return {
        "stop": "end_turn",
        "length": "max_tokens",
        "tool_calls": "tool_use",
        "content_filter": "stop_sequence",
    }.get(str(reason), str(reason))

File: gobackend/internal/server/anthropic.go
package server

import (
	"encoding/json"
	"fmt"
	"strings"
)

// anthropicToOpenAI converts an Anthropic Messages request into an OpenAI chat
// request, preserving block ordering. Mirrors qb2api/anthropic.py.
func anthropicToOpenAI(body map[string]any) (map[string]any, error) {
	if body == nil {
		return nil, fmt.Errorf("request body must be an object")
	}
	messages := []any{}
	if system := contentToText(body["system"]); system != "" {
		messages = append(messages, map[string]any{"role": "system", "content": system})
	}
	rawMessages, _ := body["messages"].([]any)
	for _, raw := range rawMessages {
		message, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		messages = append(messages, anthropicMessageToOpenAI(message)...)
	}
	request := map[string]any{
		"model":    body["model"],
		"messages": messages,
		"stream":   boolOr(body["stream"], false),
	}
	for key, value := range completionOptions(body) {
		request[key] = value
	}
	for key, value := range metadataOptions(body) {
		request[key] = value
	}
	for key, value := range toolOptions(body) {
		request[key] = value
	}
	for _, key := range []string{"reasoning_effort", "context_window", "max_context_tokens"} {
		if value, ok := body[key]; ok && value != nil {
			request[key] = value
		}
	}
	return request, nil
}

func completionOptions(body map[string]any) map[string]any {
	out := map[string]any{}
	if value, ok := body["max_tokens"]; ok && value != nil {
		out["max_tokens"] = value
	}
	for _, pair := range [][2]string{{"temperature", "temperature"}, {"top_p", "top_p"}, {"stop_sequences", "stop"}} {
		if value, ok := body[pair[0]]; ok && value != nil {
			out[pair[1]] = value
		}
	}
	return out
}

func metadataOptions(body map[string]any) map[string]any {
	metadata, ok := body["metadata"].(map[string]any)
	if !ok {
		return nil
	}
	if user, ok := metadata["user_id"].(string); ok && user != "" {
		return map[string]any{"user": user}
	}
	return nil
}

func toolOptions(body map[string]any) map[string]any {
	out := map[string]any{}
	rawTools, _ := body["tools"].([]any)
	tools := make([]any, 0, len(rawTools))
	for _, raw := range rawTools {
		tool, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		tools = append(tools, map[string]any{
			"type": "function",
			"function": map[string]any{
				"name":        stringOr(tool["name"], ""),
				"description": stringOr(tool["description"], ""),
				"parameters":  mapOr(tool["input_schema"], map[string]any{"type": "object", "properties": map[string]any{}}),
			},
		})
	}
	if len(tools) > 0 {
		out["tools"] = tools
	}
	if choice := anthropicToolChoice(body["tool_choice"]); choice != nil {
		out["tool_choice"] = choice
	}
	return out
}

func anthropicToolChoice(raw any) any {
	switch typed := raw.(type) {
	case nil:
		return nil
	case string:
		return typed
	case map[string]any:
		switch stringOr(typed["type"], "") {
		case "auto":
			return "auto"
		case "any":
			return "required"
		case "none":
			return "none"
		case "tool":
			return map[string]any{"type": "function", "function": map[string]any{"name": stringOr(typed["name"], "")}}
		}
		return typed
	}
	return raw
}

// anthropicMessageToOpenAI converts one Anthropic message into one or more
// OpenAI messages, splitting tool_result blocks into tool turns.
func anthropicMessageToOpenAI(message map[string]any) []any {
	role := stringOr(message["role"], "user")
	content := message["content"]
	if role == "assistant" {
		if blocks, ok := content.([]any); ok {
			return []any{assistantMessage(blocks)}
		}
	}
	if text, ok := content.(string); ok {
		return []any{map[string]any{"role": role, "content": text}}
	}
	blocks, ok := content.([]any)
	if !ok {
		return []any{map[string]any{"role": role, "content": contentToText(content)}}
	}
	out := []any{}
	textParts := []string{}
	for _, raw := range blocks {
		block, ok := raw.(map[string]any)
		if !ok {
			textParts = append(textParts, contentToText(raw))
			continue
		}
		switch stringOr(block["type"], "") {
		case "text":
			textParts = append(textParts, stringOr(block["text"], ""))
		case "tool_result":
			if len(textParts) > 0 {
				out = append(out, map[string]any{"role": role, "content": strings.Join(nonEmpty(textParts), "\n")})
				textParts = textParts[:0]
			}
			toolCallID := stringOr(block["tool_use_id"], stringOr(block["id"], ""))
			out = append(out, map[string]any{
				"role":         "tool",
				"tool_call_id": toolCallID,
				"content":      contentToText(block["content"]),
			})
		case "image":
			textParts = append(textParts, "[image]")
		default:
			textParts = append(textParts, contentToText(block))
		}
	}
	if len(textParts) > 0 {
		out = append(out, map[string]any{"role": role, "content": strings.Join(nonEmpty(textParts), "\n")})
	}
	if len(out) == 0 {
		return []any{map[string]any{"role": role, "content": ""}}
	}
	return out
}

func assistantMessage(blocks []any) map[string]any {
	textParts := []string{}
	toolCalls := []any{}
	for _, raw := range blocks {
		block, ok := raw.(map[string]any)
		if !ok {
			textParts = append(textParts, contentToText(raw))
			continue
		}
		switch stringOr(block["type"], "") {
		case "text":
			textParts = append(textParts, stringOr(block["text"], ""))
		case "tool_use":
			arguments, _ := json.Marshal(mapOr(block["input"], map[string]any{}))
			id := stringOr(block["id"], "")
			if id == "" {
				id = "toolu_" + randomToken(12)
			}
			toolCalls = append(toolCalls, map[string]any{
				"id":   id,
				"type": "function",
				"function": map[string]any{
					"name":      stringOr(block["name"], ""),
					"arguments": string(arguments),
				},
			})
		}
	}
	message := map[string]any{
		"role":    "assistant",
		"content": strings.Join(nonEmpty(textParts), "\n"),
	}
	if len(toolCalls) > 0 {
		message["tool_calls"] = toolCalls
	}
	return message
}

// openAIToAnthropic converts a non-streaming OpenAI completion into an
// Anthropic message response.
func openAIToAnthropic(response map[string]any, model string) map[string]any {
	choices, _ := response["choices"].([]any)
	choice := map[string]any{}
	if len(choices) > 0 {
		choice, _ = choices[0].(map[string]any)
	}
	message, _ := choice["message"].(map[string]any)
	if message == nil {
		message = map[string]any{}
	}
	content := []any{}
	if thinking := stringOr(message["reasoning_content"], ""); thinking != "" {
		content = append(content, map[string]any{"type": "thinking", "thinking": thinking, "signature": ""})
	}
	if text := stringOr(message["content"], ""); text != "" {
		content = append(content, map[string]any{"type": "text", "text": text})
	}
	if calls, ok := message["tool_calls"].([]any); ok {
		for _, raw := range calls {
			call, ok := raw.(map[string]any)
			if !ok {
				continue
			}
			function, _ := call["function"].(map[string]any)
			id := stringOr(call["id"], "")
			if id == "" {
				id = "toolu_" + randomToken(16)
			}
			content = append(content, map[string]any{
				"type":  "tool_use",
				"id":    id,
				"name":  stringOr(function["name"], ""),
				"input": parseToolInput(function["arguments"]),
			})
		}
	}
	usage, _ := response["usage"].(map[string]any)
	return map[string]any{
		"id":            anthropicMessageID(response["id"]),
		"type":          "message",
		"role":          "assistant",
		"model":         model,
		"content":       content,
		"stop_reason":   mapFinishReason(choice["finish_reason"]),
		"stop_sequence": nil,
		"usage": map[string]any{
			"input_tokens":  intFrom(usage, "prompt_tokens", "input_tokens"),
			"output_tokens": intFrom(usage, "completion_tokens", "output_tokens"),
		},
	}
}

func anthropicMessageID(raw any) string {
	id := stringOr(raw, "")
	if strings.HasPrefix(id, "msg_") {
		return id
	}
	suffix := strings.TrimPrefix(id, "chatcmpl-")
	if suffix == "" {
		suffix = randomToken(16)
	}
	return "msg_" + suffix
}

func mapFinishReason(reason any) any {
	text, ok := reason.(string)
	if !ok || text == "" {
		return nil
	}
	switch text {
	case "stop":
		return "end_turn"
	case "length":
		return "max_tokens"
	case "tool_calls":
		return "tool_use"
	case "content_filter":
		return "stop_sequence"
	}
	return text
}

func parseToolInput(raw any) any {
	switch typed := raw.(type) {
	case map[string]any:
		return typed
	case string:
		if typed == "" {
			return map[string]any{}
		}
		var parsed any
		if err := json.Unmarshal([]byte(typed), &parsed); err != nil {
			return map[string]any{"_raw": typed}
		}
		if mapping, ok := parsed.(map[string]any); ok {
			return mapping
		}
		return map[string]any{"value": parsed}
	}
	return map[string]any{}
}

// contentToText flattens an Anthropic content value into plain text.
func contentToText(content any) string {
	switch typed := content.(type) {
	case nil:
		return ""
	case string:
		return typed
	case []any:
		parts := []string{}
		for _, item := range typed {
			mapping, ok := item.(map[string]any)
			if !ok {
				parts = append(parts, contentToText(item))
				continue
			}
			switch stringOr(mapping["type"], "") {
			case "text":
				parts = append(parts, stringOr(mapping["text"], ""))
			case "tool_result":
				parts = append(parts, contentToText(mapping["content"]))
			case "image":
				parts = append(parts, "[image]")
			default:
				encoded, _ := json.Marshal(mapping)
				parts = append(parts, string(encoded))
			}
		}
		return strings.Join(nonEmpty(parts), "\n")
	case map[string]any:
		if stringOr(typed["type"], "") == "text" {
			return stringOr(typed["text"], "")
		}
		encoded, _ := json.Marshal(typed)
		return string(encoded)
	}
	return fmt.Sprint(content)
}

func nonEmpty(values []string) []string {
	out := make([]string, 0, len(values))
	for _, value := range values {
		if value != "" {
			out = append(out, value)
		}
	}
	return out
}

func boolOr(value any, fallback bool) bool {
	if flag, ok := value.(bool); ok {
		return flag
	}
	return fallback
}

func stringOr(value any, fallback string) string {
	if text, ok := value.(string); ok {
		return text
	}
	return fallback
}

func mapOr(value any, fallback map[string]any) map[string]any {
	if mapping, ok := value.(map[string]any); ok {
		return mapping
	}
	return fallback
}

func intFrom(usage map[string]any, names ...string) int {
	for _, name := range names {
		switch value := usage[name].(type) {
		case float64:
			return int(value)
		case int:
			return value
		}
	}
	return 0
}
